products/libllbuild/BuildDB-C-API.cpp

```diff
@@ -0,0 +1,286 @@
+  
+  const bool buildStarted(std::string *error_out) {
+    return _db.get()->buildStarted(error_out);
+  }
+  
+  void buildComplete() {
+    _db.get()->buildComplete();
+  }
+  
+  const bool getKeys(std::map<KeyID, KeyType>& keys_out, std::string *error_out) {
+    return _db.get()->getKeys(keys_out, error_out);
+  }
+  
+  void dump() {
+    _db.get()->dump(llvm::outs());
+  }
+};
+
+}
+
+const llb_data_t mapData(std::vector<uint8_t> input) {
+  const auto data = input.data();
+  const auto size = sizeof(data) * input.size();
+  const auto newData = (uint8_t *)malloc(size);
+  memcpy(newData, data, size);
+  
+  return llb_data_t {
+    input.size(),
+    newData
+  };
+}
+
+const llb_database_result_t mapResult(Result result) {
+  
+  auto count = result.dependencies.size();
+  auto data = result.dependencies.data();
+  auto size = sizeof(data) * count;
+  KeyID *deps = (KeyID*)malloc(size);
+  memcpy(deps, data, size);
+  
+  return llb_database_result_t {
+    mapData(result.value),
+    result.signature.value,
+    result.computedAt,
+    result.builtAt,
+    deps,
+    static_cast<uint32_t>(count)
+  };
+}
+
+void llb_database_destroy_result(llb_database_result_t *result) {
+  delete result->dependencies;
+}
+
+const llb_database_t* llb_database_create(
+                                    char *path,
+                                    uint32_t clientSchemaVersion,
+                                    llb_database_delegate_t delegate,
+                                    llb_data_t *error_out) {
+  assert(delegate.get_key_for_id);
+  assert(delegate.get_key_id);
+  
+  std::string error;
+  
+  auto database = new CAPIBuildDB(StringRef(path), clientSchemaVersion, delegate, &error);
+  
+  if (!error.empty() && error_out) {
+    error_out->length = error.size();
+    error_out->data = (const uint8_t*)strdup(error.c_str());
+  }
+  
+  return (llb_database_t *)database;
+}
+
+void llb_database_destroy(llb_database_t *database) {
+  delete (CAPIBuildDB *)database;
+}
+
+const uint64_t llb_database_get_current_iteration(llb_database_t *database, bool *success_out, llb_data_t *error_out) {
+  auto db = (CAPIBuildDB *)database;
+  std::string error;
+  
+  auto iteration = db->getCurrentIteration(success_out, &error);
+  
+  if (!error.empty() && error_out) {
+    error_out->length = error.size();
+    error_out->data = (const uint8_t*)strdup(error.c_str());
+  }
+  
+  return iteration;
+}
+
+void llb_database_set_current_iteration(llb_database_t *database, uint64_t value, llb_data_t *error_out) {
+  auto db = (CAPIBuildDB *)database;
+  std::string error;
+  
+  db->setCurrentIteration(value, &error);
+  
+  if (!error.empty() && error_out) {
+    error_out->length = error.size();
+    error_out->data = (const uint8_t*)strdup(error.c_str());
+  }
+}
+
+const bool llb_database_lookup_rule_result(llb_database_t *database, llb_database_key_id keyID, llb_database_key_type ruleKey, llb_database_result_t *result_out, llb_data_t *error_out) {
```

Comment:
What I have been doing recently with APIs where the memory management is awkward is instead exposing a callback API -- the C code will call a block or a function pointer with each result. In the Swift code, it is then very easy to accumulate the result, and this avoids the need to allocate memory which is likely more efficient.

Blocks are certainly more convenient than C function pointers when bridged to Swift; given how Swift focus llbuild is, I think it might be acceptable to say that llbuild requires building with a compiler which supports blocks. That is fairly readily accessible these days, on many platforms.

Reply:
(again, though, I'd propose bringing in the individual APIs after the initial PR)

Reply:
Alternately, we might want to consider if llb_data_t needs to get a bit to indicate whether or not it owns its memory. However, I'd like to avoid unnecessary copying if possible